### ISSUE TITLE
Use of "configure" is/maybe misleading

### ISSUE THREAD

**Issue by robaol** [microsoft/vscode-makefile-tools#205]
The extension and its documentation use the terms "Configure" and "Configuration" a lot.

This is really confusing/misleading when dealing with makefiles and autotools where ./configure has a very specific and significant meaning.

The actions menu contains
Pre-Configure
Configure
Clean Configure
Build clean
What are these going to do? 
Maybe using the word "make" with the target, e.g. "make clean" instead of "Build clean" would be clearer. I cant guess what Configure is trying to do? Maybe it should be called "Run ./configure"? I don't really want to run ./configure by accident, especially from the wrong folder.

For me, every time I start vscode, the extension says it is generating configuration cache. This doesn't ever seem to finish, even after 3+ minutes of the progress bar having completed. When i cancel it, it then says "Configure failed". So, I think maybe it _is_ trying to run ./configure every time, which I don't want.

I am building "Hamlib". I create a build folder, change into it and run 
mkdir build && cd build
../hamlib/bootstrap
../hamlib/configure
make

This keeps intermediate files and build product out of the source tree and makes it easier to keep the repo clean.

The system is (uname -a):
Linux raspberrypi 5.10.52-v7l+ #1441 SMP Tue Aug 3 18:11:56 BST 2021 armv7l GNU/Linux

**Comment by bobbrow** [comment]
The extension does not attempt to run `./configure` scripts for you unless you tell us to in the settings.  The "configure" we are referring to is configuring IntelliSense.  To do this we need the compilation commands for every source file in the project, so we do a "dry run" of your makefile to generate logs which we then parse for compilation information.

**Comment by robaol** [comment]
@bobbrow Thanks, I will try it out. I am able to make the project ok but dont know how to collect the output - I plan to examine the extension's dry run flags as a start. I'll let you know how I get on.

**Comment by andreeis** [comment]
@robaol, if you can share your project with us (or a reduced version that manifests the problem) we can do all the investigation and answer you how to best overcome the long dryrun problem.

**Comment by robaol** [comment]
@andreeis thank you for following up on this.

I am building [Hamlib](https://github.com/robaol/Hamlib), which is a computer control interface abstraction library for radios. 

There are some install pre-req [packages](https://github.com/robaol/Hamlib/blob/71fc5534b74d9a66777269e7dad6acb6b3b2f5b2/README.developer#L231) (I expect you have the tooling ones). I had to install the following on raspberry pi-os (I think the names can vary by distro):
autoconf
libtool
libtool-bin
libreadline-dev
libusb-1.0.0-dev

I am using the out of tree building instructions in [README.developer](https://github.com/robaol/Hamlib/blob/71fc5534b74d9a66777269e7dad6acb6b3b2f5b2/README.developer#L299). I don't run the make install and instead have edited LD_LIBRARY_PATH  in my launch.json so that I can debug/run one of the test harness programs. I doubt you will want to run the program but please look in my launch.json config on my Add_icm710itu branch.

Anyway, the ../hamlib/bootstrap and ../hamlib/configure work fine. But when I select the makefile extension, it starts a dry-run and never finishes. 
My settings.json is:
`{
    "makefile.extensionOutputFolder": ".vscode",
    "makefile.buildLog": "",
    "makefile.configurationCachePath": "../vscode/configurationCache.log",
    "makefile.makeDirectory": "../build",
    "makefile.configurations": [
           
    ]
}
`
(Insert code md didnt work for the above)

I have put the configurationCachePath outside of the source tree as I don't want to commit it by accident. I guess I should have put it in the build tree - I could not find much best practice to follow.

Please let me know if need anything more. I look forward to hearing from you.
Regards
Rob

**Comment by robaol** [comment]
> For what is worth, I also find those names are confusing in this context.
> I'd suggest "Configure IntelliSense", I also though that "Configure" would run `./configure`.
> Looks like I'd use "Pre-Configure" for `./configure`, did I miss where the other commands are explained in the docs?
> "Clean Configure" cleans the IntelliSense configuration?

Any more thoughts about making the extension commands more explicit in saying that it is intellisense that is being configured or, that a make file target is going to be run?
Unfortunately, "configure" and "configuration" are massively over-loaded terms in software. I think a bit of disambiguation would be an improvement...

Regards
Rob

**Comment by andreeis** [comment]
@robaol, I agree about the terminology, thank you for making the good points. We will discuss over here.

It's great that you were able to share your project. I will have a look. We are focusing on a different project this week and I may not be able to try your repro but there is something you can do to help us answer faster.

Write down the sequence of commands that you normally run in your developer prompt (you said bootstrap, configure... then you would run "make" but with what params?).

Then, set "makefile.loggingLevel": "Debug" and "makefile.extensionLog": "extension.log" in your .vscode/settings.json. Load your project, let is run for 2 min and share with us .vscode\extension.log and .vscode\dryrun.log.

I may spot something by reading these files without cloning and configuring the project end to end.

Then, to get unblocked even before I have a hypothesis for you, make a full clean build of your project in your developer terminal, save the log (and try to not have "make" being silenced via directives), set "makefile.buildLog": "your/path/to/the/full/build.log" in .vscode/settings.json and load the project again. This should skip over running make --dry-run, should be instantaneous and you should have IntelliSense and all the other features (unless there are parser bugs or complicated known scenarios in how your build reports the logging which we'd also be interested in to improve).

**Comment by andreeis** [comment]
@robaol, I cloned and tested the Hamlib repository. I confirm that this project is in the unfortunate category of projects which cannot benefit of --always-make switch, because it conflicts with configure targets causing them to run indefinitely.

What you can do is:
- as I said in my reply above, you can manually save the output of a full clean build of your project and point the extension to it via "makefile.buildLog". Refresh every time you make a change relevant for IntelliSense (add/remove c/c++ file, add/remove/change compiler switch)
- if you want to rely on a "live" --dry-run as opposed to a "frozen" build.log, you have to remove the --always-make from "makefile.dryrunSwitches". The major downside to this is that make will not list any compilation commands for source files that are "up to date" and you will not have IntelliSense information for those files (unless they are present in the configurationCache, meaning that you did at least one configure step when your project was clean). To mitigate this you can try the following:
      1. Run make clean, load your project in VSCode and let is configure and then don't ever run "Clean Configure" unless your project is also clean. The following simple (not clean) configure steps, while --always-make is missing, will not see any up to date source files but they should read that from the IntelliSense cache we create or next time you edit the source files or the build steps that impact them (whatever makes the "make" tool compile it again).
      2. I experimented once with the "touch" concept of "make". It was interesting and solved some scenarios but I didn't have time to properly elaborate that for more projects, there were some issues. You may experiment and figure out a way to replace --always-make by a "touch" step that would make all the targets you are interested in being included for IntelliSense as "out of date". Once you have such command crafted and tested you could share it with us because we can implement such approach in the extension to help the developers that can't use --always-make switch.

**Comment by robaol** [comment]
@andreeis Thank you for investigating and testing this. I'm sorry that work has kept me busy and I have not had spare time for this project recently or to try your suggestions.
I hope to get back to this project soon (Xmas holidays :) ) and will try out your suggestions and see which work best. If I make any useful commands, I will share them. 
Do you have any links on how to start writing/modifying commands that you could share, for me to refer to?
Thanks again.

**Comment by andreeis** [comment]
@robaol, I was experimenting with the "--touch" or "-t" switch of "make" so that I force "make" to see and build targets even if they would normally be "up-to-date", even if "--always-make" wouldn't be passed to "make". But I think in my experiments I wasn't "touching" the right things.

I just found now the "touch" command on linux and it appears that the equivalent can be achieved [on windows](https://superuser.com/questions/10426/windows-equivalent-of-the-linux-command-touch/764716). I think sometime soon we will incorporate this into the extension so that when users open a c/c++ file that we think we don't have IntelliSense for, if we see "--always-make" is missing, we can run this "touch"-ing command on that file and invoke the "--dry-run" again. This will give IntelliSense for the files that are not normally "seen" by "make".

What you can do in the meantime, write a script that would "touch" all the c/c++ source files that you are interested in. You can hook that script to the extension via "makefile.preConfigureScript" and make sure to run the command "Makefile: Preconfigure" from the palette before you configure (there is also the boolean setting "makefile.alwaysPreConfigure", it can be handy in your case).

The downside of this will be that you won't be able to build incrementally again. Once you "touch" these files, they will be seen out of date. Next time you do a real build, it won't be as fast, it would be like you would have written code in all of these files. Unfortunately, I didn't find a way for "make" to see something for --dry-run only and not see it again later for a real build. Unless you also "touch" the results of the build (objs, libs, exes....etc... complicated and error prone) if you know for sure which is really up-to-date still.
